test/jdk/com/sun/jdi/JdwpOnThrowTest.java

```diff
@@ -0,0 +1,153 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/*
+ * @test
+ * @bug 8317920
+ * @summary Tests for JDWP agent to send valid exception event with onthrow option
+ * @library /test/lib
+ *
+ * @build ThrowCaughtException JdwpOnThrowTest
+ * @run main/othervm JdwpOnThrowTest
+ */
+public class JdwpOnThrowTest {
+
+    private static long TIMEOUT = 10000;
+
+    private static String ATTACH_CONNECTOR = "com.sun.jdi.SocketAttach";
+    // cache socket attaching connector
+    private static AttachingConnector attachingConnector;
+
+    public static void main(String[] args) throws Exception {
+        int port = findFreePort();
+        try (Debuggee debuggee = Debuggee.launcher("ThrowCaughtException").setAddress("localhost:" + port)
+                                         .enableOnThrow("Ex", "Start").setSuspended(true).launch()) {
+            VirtualMachine vm = null;
+            try {
+                vm = attach("localhost", "" + port);
+                EventQueue queue = vm.eventQueue();
+                log("Waiting for exception event");
+                long start = System.currentTimeMillis();
+                while (start + TIMEOUT > System.currentTimeMillis()) {
+                    EventSet eventSet = queue.remove(TIMEOUT);
+                    EventIterator eventIterator = eventSet.eventIterator();
+                    while(eventIterator.hasNext() && start + TIMEOUT > System.currentTimeMillis()) {
+                        Event event = eventIterator.next();
+                        if (event instanceof ExceptionEvent ex) {
+                            if (ex.exception() == null) {
+                                throw new RuntimeException("Exception is null");
+                            }
+                            if (ex.exception().type() == null) {
+                                throw new RuntimeException("Exception type is null");
+                            }
+                            if (ex.exception().referenceType() == null) {
+                                throw new RuntimeException("Exception reference type is null");
+                            }
+                            if (ex.catchLocation() == null) {
+                                throw new RuntimeException("Exception catch location is null");
+                            }
+                            if (!ex.location().equals(ex.thread().frame(0).location())) {
+                                throw new RuntimeException(
+                                    String.format("Throw location %s and location of first frame %s are not equal",
+                                                  ex.location(), ex.thread().frame(0).location()));
+                            }
+                            if (!ex.exception().type().name().equals("Ex")) {
+                                throw new RuntimeException("Exception has wrong type: " + ex.exception().type().name());
+                            }
```

Comment:
Not a review but just a suggestion to refactor a little bit:
The lines 75-94 can be placed into a method `verifyExceptionEvent(ExceptionEvent ex)`,
so the `main()` method is smaller and better observable.

Reply:
I was also thinking something similar.